gradle/spec/dependabot/gradle/update_checker/version_finder_spec.rb

```diff
@@ -694,6 +694,34 @@
           )
       end
 
+      context "with credentials" do
+        let(:credentials) do
+          [{
+            "type" => "maven_repository",
+            "url" => "https://dl.bintray.com/magnusja/maven/",
+            "username" => "username",
+            "password" => "password"
+          }]
+        end
+
+        before do
+          stub_request(:get, google_metadata_url).
+            to_return(status: 404, body: "")
+          stub_request(:get, magnusja_metadata_url).
+            with(basic_auth: %w(username password)).
+            to_return(
+              status: 200,
+              body: fixture("google_metadata", "com_google_guava.xml")
+            )
+        end
+
+        it "calls custom repo using defined credentials" do
+          subject
```

Comment:
I don't really understand how this setup hits the newly added code 😬 can you elaborate a little? I'm wondering if we can structure this in a way to make it more clear what we're testing. Without the check in place would we end up making the request twice?

Reply:
Exactly right, a private repo would be called twice.

I actually did not get the issue exactly right initially, because there is another one hiding in the specific usecase when using mapbox repo from my example. I only got the exact problem when writing the test 😄 

What happens now, the private repo would be first called without any auth headers (since it is not filtered out and it's parsed from the build file) and get 401 unauthorised which would put it in to `forbidden_urls` object. Eventually it gets called again with auth headers, but the endpoint core constructs to fetch metadata, like `com/mapbox/search/mapbox-search-android-ui/maven-metadata.xml`, doesn't exist on mapbox private repo, so it returns 404, not found.

Because there was a first call which returned 401, check [version_details.none? && forbidden_urls.any?](https://github.com/dependabot/dependabot-core/blob/b78e5d7ca40d6f51aaa41339067533802e161612/gradle/lib/dependabot/gradle/update_checker/version_finder.rb#L68) returns `true` and authentication failure is raised which is suuuper confusing 😅 

So without knowing these specifics, it's quite hard to describe what this test is actually checking. It would be kind of obvious if we simply validate the `repositories` array explicitly, but calling a private method is no bueno. And since technically, in case of normal usecase, the end result is not really affected, validating subjects return values wouldn't actually properly test this either.

I'm definitely open to suggestion on how to better structure the spec or maybe some better naming of context and spec?